### PR TITLE
[IR] Refactor `JvmSymbols` for on-demand `java.lang.reflect` symbols

### PR DESCRIPTION
**Pull request by kandersen**
Dependent on #4599

### CODE REVIEW

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/JvmReflectSymbols.kt [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+                    name = Name.identifier("args")
+                    type = irBuiltIns.arrayClass.typeWith(irBuiltIns.anyNType)
+                    varargElementType = irBuiltIns.anyNType
+                }
+            }
+        }
+
+    init {
+        val klass = context.ir.symbols.javaLangClass.owner
+        if (!klass.functions.any { it.name.asString() == "getDeclaredMethod" }) {
```
Comment (kandersen) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/JvmReflectSymbols.kt:80]: @udalov Is this "too defensive"? What I am thinking is by doing this mutation of JvmSymbols, I guess it's possible to create an inconsistent state if `JvmReflectSymbols` is instantiated a number of times.

Comment (kandersen) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/JvmReflectSymbols.kt]: Nevermind, figured lazy initialization was a cleaner, clearer way of guarding for the same thing.

**Comment by kandersen** [comment]
@udalov rebased on recent master.

File: .idea/artifacts/kotlinx_cli_jvm_1_6_255_SNAPSHOT.xml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+<component name="ArtifactManager">
+  <artifact type="jar" name="kotlinx.cli-jvm-1.6.255-SNAPSHOT">
+    <output-path>$PROJECT_DIR$/kotlin-native/endorsedLibraries/kotlinx.cli/build/libs</output-path>
+    <root id="archive" name="kotlinx.cli-jvm-1.6.255-SNAPSHOT.jar">
+      <element id="module-output" name="kotlin.kotlin-native.endorsedLibraries.kotlinx.cli.jvmMain" />
+    </root>
+  </artifact>
+</component>
```
Comment (udalov) [inline .idea/artifacts/kotlinx_cli_jvm_1_6_255_SNAPSHOT.xml:8]: Probably committed by accident?

Comment (kandersen) [reply .idea/artifacts/kotlinx_cli_jvm_1_6_255_SNAPSHOT.xml]: Absolutely! Removed.

**Comment by udalov** [comment]
Thank you!

**Pull request merged**